### PR TITLE
Filter Implementation 

### PR DESCRIPTION
**Pull request by ani-sha**
Initial Version - Addition of filtering capabilities #16, few cases are yet to be validated.

### CODE REVIEW

File: engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java [hunk 1]
```diff
@@ -263,25 +251,138 @@ public String getSQL() {
...
             c = left.eq(((BooleanValue)v).isValue());
-        } else if (v instanceof IntValue) {
+        else if(conditionName.equals("ne"))
+            c = left.ne(((BooleanValue) v).isValue());
+        return c;
+    }
+
+    public Condition visitIntValue(Value v, org.jooq.Field left, String conditionName) {
+        Condition c = null;
+        if(conditionName.equals("eq"))
```
Comment (wtrocki) [inline engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java:306]: Amazing job! I see you have all arguments covered. Is that ready for verification?

Comment (ani-sha) [reply engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java:306]: Have to implement `in` and `between` for `float` and `int` types.

File: engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java [hunk 2]
```diff
@@ -195,6 +195,179 @@ public void testDeepNestedQuery() throws Exception {
...
+                "    select json_arrayagg(json_object(key 'ACCOUNT_ID' value \"g1\".\"ACCOUNT_ID\"))\n" +
+                "    from PUBLIC.ACCOUNT \"g1\"\n" +
+                "    where (\n" +
+                "      \"g0\".\"SSN\" = \"g1\".\"SSN\"\n" +
+                "      and \"g1\".\"ACCOUNT_ID\" = 19980001\n" +
+                "    )\n" +
+                "  ) \"accounts\"\n" +
+                "from PUBLIC.CUSTOMER \"g0\"\n" +
+                "where (\n" +
+                "  (\n" +
```
Comment (rareddy) [inline engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java:360]: The result of this is not correct. it should be like ```lastname = 'smith' and firstname='John' and ssn = 'cst01002'``` a `or` as enclosing does not mean the attributes inside are separated by  `or`. `or` is always adjunct operation where you could have said

```
{
  customers (filter: {
    LASTNAME: {
      eq: "Smith"
    },
    FIRSTNAME: {
      eq: "John"
    },
    or: {
      SSN: {
        eq: "CST01002"
      }
    }
  }
}
```

which should result in `(lastname = 'smith' and firstname='John' ) or ssn = 'cst01002'` 

overall what I am saying `or` should not have had any effect on the query above

File: engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java [hunk 3]
```diff
@@ -195,6 +195,179 @@ public void testDeepNestedQuery() throws Exception {
...
+                "  \"g0\".\"SSN\" \"SSN\",\n" +
+                "  (\n" +
+                "    select json_arrayagg(json_object(key 'ACCOUNT_ID' value \"g1\".\"ACCOUNT_ID\"))\n" +
+                "    from PUBLIC.ACCOUNT \"g1\"\n" +
+                "    where \"g0\".\"SSN\" = \"g1\".\"SSN\"\n" +
+                "  ) \"accounts\"\n" +
+                "from PUBLIC.CUSTOMER \"g0\"\n" +
+                "where (\n" +
+                "  \"g0\".\"LASTNAME\" like ('%' || replace(\n" +
+                "    replace(\n" +
```
Comment (rareddy) [inline engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java:253]: is JOOQ doing this conversion? seems cryptic, most databases have `startsWith` and `contains` kind of functions.

File: engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java [hunk 4]
```diff
@@ -200,7 +188,7 @@ public void startVisitRootObject(Field rootField, GraphQLFieldDefinition rootDef
     }
 
     @Override
-    public void endVisitRootObject(Field rootFeild, GraphQLFieldDefinition rootDefinition, GraphQLObjectType type) {
+    public void endVisitRootObject(Field rootField, GraphQLFieldDefinition rootDefinition, GraphQLObjectType type) {
```
Comment (rareddy) [inline engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java:191]: I need to learn how to spell `field`, I make this mistake quite often :)

File: engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java (hunk 2 above)
Comment (ani-sha) [reply engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java:360]: Thanks for pointing out that, I will correct it.

File: engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java (hunk 3 above)
Comment (ani-sha) [reply engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java:253]: Agree, it's JOOQ. I didn't understand it as well.

File: engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java (hunk 2 above)
Comment (machi1990) [reply engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java:360]: We found ourselves looking at the semantics of `or` operator recently in Graphback

/cc @craicoverflow if you want to share your experience.

Comment (craicoverflow) [reply engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java:360]: yeah, We checked out competitors (AppSync) and their query output match that by @rareddy above.

File: engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java [hunk 5]
```diff
@@ -271,67 +273,130 @@ public void visitArgument(Field field, GraphQLFieldDefinition definition, GraphQ
 
     public Condition visitStringValue(Value v, org.jooq.Field left, String conditionName) {
```
Comment (rareddy) [inline engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java:274]: Change the parameter `Value v' to `String` and remove the class castings below. When you are calling this method you already know the type just cast once there. Same for others below too

File: engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java [hunk 6]
```diff
@@ -271,67 +273,130 @@ public void visitArgument(Field field, GraphQLFieldDefinition definition, GraphQ
...
+        }
+        return c;
+    }
+
+
+    public Condition visitArrayValue(List<Value> v, org.jooq.Field left, String conditionName) {
+        Condition c = null;
+        switch (conditionName) {
+            case "between":
+                if (v.get(0) instanceof FloatValue)
```
Comment (rareddy) [inline engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java:376]: even they are single line `if` or `for` routines wrap them with curly `{` braces, makes it easier to read the code

File: engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java [hunk 7]
```diff
@@ -271,67 +273,130 @@ public void visitArgument(Field field, GraphQLFieldDefinition definition, GraphQ
...
+                    List<BigDecimal> list = new ArrayList<>();
+                    for (Value value : v)
+                        list.add(((FloatValue) value).getValue());
+                    c = left.in(list);
+                } else if (v.get(0) instanceof IntValue) {
+                    List<BigInteger> list = new ArrayList<>();
+                    for (Value value : v)
+                        list.add(((IntValue) value).getValue());
+                    c = left.in(list);
+                }
```
Comment (rareddy) [inline engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java:397]: put in a `default` in here and throw a `runtimeexception` saying wrong condition

File: engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java [hunk 8]
```diff
@@ -271,67 +273,130 @@ public void visitArgument(Field field, GraphQLFieldDefinition definition, GraphQ
...
+                break;
+            case "le":
+                c = left.le(((FloatValue) v).getValue());
+                break;
+            case "gt":
+                c = left.gt(((FloatValue) v).getValue());
+                break;
+            case "ge":
+                c = left.ge(((FloatValue) v).getValue());
+                break;
```
Comment (rareddy) [inline engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java:366]: put in `default` here and throw a `RuntimeException` saying wrong condition same at all other similar places.

File: engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java [hunk 9]
```diff
@@ -362,12 +429,15 @@ public void visitFilterInputs(Value argumentValue, VistorContext vistorContext,
                     left = field(name(vistorContext.alias, field.getName()));
                     org.jooq.Field<Object> finalLeft = left;
                     String finalConditionName = conditionName;
-                    field.getValue().getChildren().forEach(child -> {
-                        Condition condition = visitValueType(((ObjectField) child).getValue(), finalLeft, ((ObjectField) child).getName());
-                        visitCondition(finalConditionName, condition);
-                    });
+                    for(Object object : field.getValue().getChildren()) {
+                        if(object instanceof ObjectField) {
```
Comment (rareddy) [inline engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java:433]: can there be any other value than `ObjectField` here? if yes capture that in else, you can throw an exception that case

Comment (ani-sha) [reply engine/src/main/java/io/graphqlcrud/SQLQueryBuilderVisitor.java:433]: Apart from `ObjectField`, we don't need to create `Condition` for others we just simply continue.

File: engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java (hunk 2 above)
Comment (rareddy) [reply engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java:421]: @ani-sha did you make sure the previous original test as is should have behaved the way I listed above, as the test is not wrong the results were

Comment (rareddy) [reply engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java:421]: @ani-sha keep the test case as is from before, make sure the results were correct.

**Review by rareddy — APPROVED** [review]
LGTM

**Pull request merged**